### PR TITLE
Fix StompFrame.copy() does not copy headers, close #7561

### LINKED ISSUE #7561: StompFrame.copy() does not copy headers
### Expected behavior
Call stompFrame.copy() do copies of content and headers.

### Actual behavior
After call stompFrame.copy() only content was copy.

### Minimal yet complete reproducer code (or URL to code)

 `StompFrame stompFrame = new DefaultStompFrame(StompCommand.CONNECT);`
 `stompFrame.headers().set(StompHeaders.HOST, "localhost");`
 `StompFrame copyFrame = stompFrame.copy();`
 `Assert.assertTrue(copyFrame.headers().isEmpty());`

### Netty version
4.1.20.Final

### JVM version (e.g. `java -version`)
java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)

### OS version (e.g. `uname -a`)
Darwin Kernel Version 16.7.0: Mon Nov 13 21:56:25 PST 2017; root:xnu-3789.72.11~1/RELEASE_X86_64 x86_64

### PR DESCRIPTION
**Pull request by amizurov**
Motivation:
  Incorrect behavior for StompFrame.copy() method.

Modification:
 Added copying of frame headers

Result:
 When you call the StompFrame.copy() method, the headers are also copied.

Fixes [#7561].

### CODE REVIEW

**Comment by Scottmitch** [comment]
thanks @amizurov ... can you sign the [ICLA as referenced in the developer guide](https://netty.io/wiki/developer-guide.html)?

File: codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompFrame.java [hunk 1]
```diff
@@ -60,7 +66,7 @@ public StompFrame retainedDuplicate() {
 
     @Override
     public StompFrame replace(ByteBuf content) {
-        return new DefaultStompFrame(command, content);
+        return new DefaultStompFrame(command, content, new DefaultStompHeaders().set(headers));
```
Comment (Scottmitch) [inline codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompFrame.java]: Please use the same strategy as https://github.com/netty/netty/commit/942b993f2b9781ff2126ff92a6be5b975dfc72ed#diff-17c17dd59c9912dd08e4dc95fa04bcab. In summary adding a `copy()` method on `DefaultStompHeaders` which leverages `DefaultHeaders#copy()`.

Comment (amizurov) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompFrame.java]: I looked to this strategy but `DefaultStompHeaders extends DefaultHeader implements StompHeaders` we can't add `StompHeaders#copy()` method it will be clashes. Maybe we can add method with name like a `StompHeaders#deepCopy()` or another way ?

Comment (Scottmitch) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompFrame.java]: We can't modify `StompHeaders` because it is an interface and that would be an API breaking change. However we can modify `DefaultStompHeaders` and do the following:

```java
    @Override
    public DefaultStompHeaders copy() {
        DefaultStompHeaders copy = new DefaultStompHeaders();
        copy.addImpl(this);
        return copy;
    }
```

We could also modify `DefaultHeaders` so we can consolidate the `copy.addImpl`, and any other work which maybe necessary in the future, but this maybe unnecessary.

Comment (amizurov) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompFrame.java]: So we can, but in `DefaultStompHeadersSubframe` used interface `StompHeaders`, how we can used implementation method maybe I misunderstood you.

Comment (amizurov) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompFrame.java]: @Scottmitch I understand that you are busy, but I would like to quickly close this defect.

**Comment by amizurov** [comment]
@Scottmitch , @normanmaurer Guys what we do with this pull request ? For a long time no activity.

File: codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompFrame.java (hunk 1 above)
Comment (Scottmitch) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompFrame.java]: > how we can access to method that not present in interface without casting or other trick? 

Unfortunately before jdk8 (which provides the `default` key word) extending interfaces is not possible without breaking API changes. So we have a few options:

1. Add a `DefaultStompHeaders copy()` method (as suggested above)
- change the member variable in `DefaultStompHeadersSubframe` so it is declared as type `DefaultStompHeaders`.
- make the new constructor in `DefaultStompHeadersSubframe` which currently takes a `StompHeaders` instead take a `DefaultStompHeaders` and make it package private
2. "break" the StompHeaders API and add a `StompHeaders copy()` method (and also override this method in `DefaultStompHeaders` as suggested above).

@normanmaurer - my guess is you would prefer the "non-breaking" API change?

**Comment by normanmaurer** [comment]
Yes

**Comment by amizurov** [comment]
Thanks a lot for your comments, please check that is all right.

File: codec-stomp/src/test/java/io/netty/handler/codec/stomp/DefaultStompFrameTest.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        AsciiString copyHeaderName = new AsciiString("foo");
+        AsciiString copyHeaderValue = new AsciiString("bar");
+        copyFrame.headers().set(copyHeaderName, copyHeaderValue);
+
+        assertFalse(sourceFrame.headers().contains(copyHeaderName, copyHeaderValue));
+        assertTrue(copyFrame.headers().contains(copyHeaderName, copyHeaderValue));
+
+        assertEquals(1, sourceFrame.headers().size());
+        assertEquals(2, copyFrame.headers().size());
+        assertNotEquals(sourceFrame.headers(), copyFrame.headers());
```
Comment (normanmaurer) [inline codec-stomp/src/test/java/io/netty/handler/codec/stomp/DefaultStompFrameTest.java]: I think you also need to call `release()` on the frames ?

Comment (amizurov) [reply codec-stomp/src/test/java/io/netty/handler/codec/stomp/DefaultStompFrameTest.java]: You mean that `release()` not affect a copied frame or that is work correct by spec ?

Comment (normanmaurer) [reply codec-stomp/src/test/java/io/netty/handler/codec/stomp/DefaultStompFrameTest.java]: Basically add:

```java
sourceFrame.release();
copyFrame.release();
```

This is needed as `DefaultStompFrame` implementations `ByteBufHolder` to ensure there are no memory leaks.

**Comment by normanmaurer** [comment]
@Scottmitch please pull in once satisfied .

File: codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompHeadersSubframe.java [hunk 3]
```diff
@@ -24,13 +24,19 @@
 
     protected final StompCommand command;
     protected DecoderResult decoderResult = DecoderResult.SUCCESS;
-    protected final StompHeaders headers = new DefaultStompHeaders();
+    protected final DefaultStompHeaders headers;
```
Comment (normanmaurer) [inline codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompHeadersSubframe.java]: @amizurov can you explain why the changes to this class are needed at all ? I think you can revert the changes

Comment (amizurov) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompHeadersSubframe.java]: So, when we create `new DefaultStompFrame` in `DefaultStompFrame #replace(...)` method we pass headers.copy() to the constructor.

Comment (normanmaurer) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompHeadersSubframe.java]: But not to this class

Comment (amizurov) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompHeadersSubframe.java]: It's super class for `DefaultStompFrame`, else it's will be look like this: 
 ```java
DefaultStompFrame(StompCommand command, ByteBuf content, DefaultStompHeaders 
headers)  {
        super(command);

        if (headers != null) {
            this.headers.set(headers);
        }

        if (content == null) {
            throw new NullPointerException("content");
        }
        this.content = content;
    }

**Comment by Scottmitch** [comment]
@normanmaurer - since you have an outstanding question I'll let you merge when satisfied :)

File: codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompHeadersSubframe.java (hunk 3 above)
Comment (normanmaurer) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/DefaultStompHeadersSubframe.java]: never mind missed it is the super-class.

**Comment by normanmaurer** [comment]
Squashed and cherry-picked into 4.1 (4afabd3f3c400109b82951e8bec95753103ea915).

@amizurov thanks a lot!
